server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -942,8 +948,37 @@ public ClusterState execute(ClusterState currentState) {
                 // for the same repository
                 final Map<String, Map<ShardId, ShardSnapshotStatus>> knownFailures = new HashMap<>();
 
+                final Version minNodeVersion = nodes.getMinNodeVersion();
+                Set<String> reposWithIncompatibleTasks = null;
+
                 for (final SnapshotsInProgress.Entry snapshot : snapshots.entries()) {
-                    if (statesToUpdate.contains(snapshot.state())) {
+                   if (reposWithIncompatibleTasks != null && reposWithIncompatibleTasks.contains(snapshot.repository())) {
+                        logger.warn("aborting snapshot [{}] because snapshot tasks for an incompatible Elasticsearch version" +
+                                " were found operating on its target repository", snapshot.snapshot());
+                        snapshotsToBwCFail.add(snapshot);
+                    } else if (statesToUpdate.contains(snapshot.state())) {
+                       if (snapshot.version().after(minNodeVersion)) {
+                           assert changedNodes : "we can only see a new older version node if the node configuration changed";
+                           // BwC path logic to deal with the situation where an older node joins the cluster after snapshots for a newer
+                           // version have been started already. In this case we can only remove the snapshot from the cluster state as
+                           // current nodes in the cluster are not be compatible with the state machine of the snapshot potentially and
+                           // could therefore corrupt the repository.
+                           // We only do this here for the in-progress snapshot states to not interfere with snapshot finalization which
+                           // we have no way to safely abort.
```

Comment:
Shouldn't we prevent an older master from even starting finalization, in case we don't remove the ready-to-finalize snapshot from the cluster state before failing over to the older master?

Similarly, I think we need to protect against failing over to an older master during a clone operation.

Reply:
I pushed fb27d1bf2f8 which deals with this point now. Just letting the logic attempt the finalization here is the easiest by far I think.